### ISSUE TITLE
A repo pruning issue: introduced by the previous commit dd5de96e8

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#38]
Compiling the following code:

```
target triple = "x86_64-pc-linux-gnu-elf"

@.str = private unnamed_addr constant [11 x i8] c"Triple.cpp\00", align 1

declare void @qux(i8*, i32)

define zeroext i1 @foo( ) #1 align 2 {
  %1 = call i32 @BAZ(i32 6)
  %2 = icmp eq i32 %1, 64
  ret i1 %2
}

define internal i32 @BAZ(i32) #1 {
  switch i32 %0, label %6 [
    i32 0, label %2
    i32 1, label %3
    i32 2, label %4
    i32 3, label %5
  ]

; <label>:2:                                      ; preds = %1
  br label %7

; <label>:3:                                      ; preds = %1, %1
  br label %7

; <label>:4:                                      ; preds = %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1
  br label %7

; <label>:5:                                      ; preds = %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1, %1
  br label %7

; <label>:6:                                      ; preds = %1
  call void @qux(i8* getelementptr inbounds ([11 x i8], [11 x i8]* @.str, i32 0, i32 0), i32 1269)
  unreachable

; <label>:7:                                      ; preds = %5, %4, %3, %2
  %8 = phi i32 [ 64, %5 ], [ 32, %4 ], [ 16, %3 ], [ 0, %2 ]
  ret i32 %8
}

; Function Attrs: nounwind uwtable
define zeroext i1 @bar() #1 align 2 {
  %1 = call i32 @BAZ(i32 16)
  %2 = icmp eq i32 %1, 16
  ret i1 %2
}

attributes #1 = { noinline }
```

Using

```
$ rm clang.db
$ opt -mtriple="x86_64-pc-linux-gnu-repo" -O3 -S ELF.ll -o Repo.ll
$ llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj Repo.ll -o Repo.o
$ opt -mtriple="x86_64-pc-linux-gnu-repo" -O3 -S ELF.ll -o Repo.2.ll
$ llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj Repo.2.ll -o Repo.2.o
```

Dumping the compilation members in the Repo.o:
```
$ repo-ticket-dump Repo.o
9084553bd6dd2ae5676791eeefdc853f
$
$ pstore-dump -compilation=9084553bd6dd2ae5676791eeefdc853f clang.db
---

- file         :
      path : clang.db
      size : 4194304
  compilations :
      - digest      : 9084553bd6dd2ae5676791eeefdc853f
        compilation :
            members :
                - digest  : 78697ba3a2140243fc8df4e8a074367c
                  fext    : { addr: 576, size: 120 }
                  name    : foo
                  linkage : external
                - digest  : 8721b572c1993789fe9e106b1083226c
                  fext    : { addr: 416, size: 120 }
                  name    : bar
                  linkage : external
                - digest  : 5ee93246ebf9e3f5fc6b822b0391f3a1
                  fext    : { addr: 832, size: 200 }
                  name    : BAZ
                  linkage : internal
                - digest  : db3492a5a944cafc992f94423ca19b28
                  fext    : { addr: 328, size: 51 }
                  name    : .str
                  linkage : internal
                - digest  : 544b3e923624bfb77bfcb80fda9170dd
                  fext    : { addr: 736, size: 56 }
                  name    : switch.table.BAZ
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\WorkingIssues\\issue1
            triple  : x86_64-pc-linux-gnu-repo

```
Dumping the compilation members in the Repo.2.o:
```
$ repo-ticket-dump Repo.2.o
60568d6ce40ef424647ed907d626de15
$
$ pstore-dump -compilation=60568d6ce40ef424647ed907d626de15 clang.db
---
---

- file         :
      path : clang.db
      size : 4194304
  compilations :
      - digest      : 60568d6ce40ef424647ed907d626de15
        compilation :
            members :
                - digest  : 78697ba3a2140243fc8df4e8a074367c
                  fext    : { addr: 544, size: 96 }
                  name    : foo
                  linkage : external
                - digest  : 8721b572c1993789fe9e106b1083226c
                  fext    : { addr: 416, size: 96 }
                  name    : bar
                  linkage : external
                - digest  : db3492a5a944cafc992f94423ca19b28
                  fext    : { addr: 328, size: 51 }
                  name    : .str
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\WorkingIssues\\issue1
            triple  : x86_64-pc-linux-gnu-repo
...
```

In the second time build (Repo.2.o), the fragments of 'BAZ' and 'switch.table.BAZ' are missed in the compilation members.

Dumping the fragment 'foo':
```
$ pstore-dump -fragment=78697ba3a2140243fc8df4e8a074367c clang.db
---

- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : 78697ba3a2140243fc8df4e8a074367c
        fragment :
            - type     : text
              contents :
                  align   : 16
                  data    :
                      - 0:      pushq   %rax
                      - 1:      movl    $6, %edi
                      - 6:      callq   0
                      - 11:     cmpl    $64, %eax
                      - 14:     sete    %al
                      - 17:     popq    %rcx
                      - 18:     retq
                  ifixups : [ ]
                  xfixups :
                      - name   : BAZ
                        type   : 4
                        offset : 7
                        addend : 18446744073709551612
...

```
The above fragment has a xfixup to the 'BAZ', which is missed in the compilation Repo.2.o. During the link stage, an error of undefined reference to 'BAZ' will be given.

**Comment by MaggieYingYi** [comment]
Analysis:

Dumping the 'BAZ' fragment:
```
$ pstore-dump -fragment=5ee93246ebf9e3f5fc6b822b0391f3a1 clang.db
---

- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : 5ee93246ebf9e3f5fc6b822b0391f3a1
        fragment :
            - type     : text
              contents :
                  align   : 16
                  data    :
                      - 0:      pushq   %rax
                      - 1:      cmpl    $4, %edi
                      - 4:      jae     12
                      - 6:      movslq  %edi, %rax
                      - 9:      movl    (,%rax,4), %eax
                      - 16:     popq    %rcx
                      - 17:     retq
                      - 18:     movl    $0, %edi
                      - 23:     movl    $1269, %esi
                      - 28:     callq   0
                  ifixups : [ ]
                  xfixups :
                      - name   : switch.table.BAZ
                        type   : 11
                        offset : 12
                        addend : 0
                      - name   : .str
                        type   : 10
                        offset : 19
                        addend : 0
                      - name   : qux
                        type   : 4
                        offset : 29
                        addend : 18446744073709551612
            - type     : dependent
              contents :
                  - digest  : 544b3e923624bfb77bfcb80fda9170dd
                    fext    : { addr: 672, size: 56 }
                    name    : switch.table.BAZ
                    linkage : internal
...
```
The 'BAZ' digest is changed in the backend since it has a dependent 'switch.table.BAZ'. During the second time build, the 'BAZ' digest is calculated by the repo metadata generation and is different from the 'BAZ' digest stored in the database. Therefore, the 'BAZ’ isn’t pruned. The function 'foo' and 'bar' don't have the dependents and they are pruned during the second time build. Since 1) the 'BAZ' is private function and 2) both 'foo' and 'bar' are changed to available_externally type after pruning, the 'BAZ' is removed by the LLVM passes.

**Comment by MaggieYingYi** [comment]
The proposed solution:

The fragment's digest (e.g. 'foo' or 'bar') need to be updated if the fragment has a relocation to another fragment (e.g. 'BAZ') whose digest is changed in the LLVM backend.

**Comment by MaggieYingYi** [comment]
Although the proposed solution (mentioned in https://github.com/SNSystems/llvm-project-prepo/issues/38#issuecomment-511894648) solves the issue, but it prevents the repo pruning since the fragment's digests (e.g. 'foo' and 'bar') change.

To solve the issue, the ‘BAZ’ need to be added into the dependents of ‘foo’ and ‘bar’.

**Comment by paulhuggett** [comment]
> The fragment's digest (e.g. 'foo' or 'bar') need to be updated if the fragment has a relocation to another fragment (e.g. 'BAZ') whose digest is changed in the LLVM backend.

Just to put into writing what we discussed on the phone: we shouldn't ever be modifying the fragment digests in the backend. If we do this, then the front-end pruning pass won't ever be able to find a match for the digest.

**Comment by MaggieYingYi** [comment]
> 
> Just to put into writing what we discussed on the phone: we shouldn't be modifying the fragment digests in the backend. If we do this, then the front-end pruning pass won't ever be able to find a match for the digest and this is critical for achieving the compile-time benefits that the repository promised..

Thanks for reminding me on this discussion. In order to fix the issue https://github.com/SNSystems/llvm-project-prepo/issues/17, we have modified the digest in the backend. The change has been committed in dd5de96e. As mentioned in the https://github.com/SNSystems/llvm-project-prepo/issues/17#issuecomment-487004398, we must include the dependents’ information in the fragment's digest since the name and digest of the dependents appears in the fragment.

The issue https://github.com/SNSystems/llvm-project-prepo/issues/38 is caused by missing ‘BAZ’ in the dependents of fragment ‘foo’ and ‘bar’. The fix is trying to add ‘BAZ’ into the dependents of fragment ‘foo’ and ‘bar’.

**Comment by paulhuggett** [comment]
> In order to fix the issue #17, we have modified the digest in the backend. The change has been committed in [dd5de96](https://github.com/SNSystems/llvm-project-prepo/commit/dd5de96e828747ccd9423162422d1aa8689d4748). As mentioned in the [#17 (comment)](https://github.com/SNSystems/llvm-project-prepo/issues/17#issuecomment-487004398), we must include the dependents’ information in the fragment's digest since the name and digest of the dependents appears in the fragment.

I think that we may need to re-work the fix for bug #17 because it has had the problematic effect of both introducing this bug, and breaking the front-end pruning pass.

**Comment by MaggieYingYi** [comment]
> 
> I think that we may need to re-work the fix for bug #17 because it has had the problematic effect of both introducing this bug, and breaking the front-end pruning pass.

After revising the issue #17, I found that issue #17 is the same as the issue #25 (a pruning issue for a global variable with zero-initializer). The commit https://github.com/SNSystems/llvm-project-prepo/commit/65d8c8d1301dbf9bbd8fb6ca0c89b94d26c35373 should fix both issue #17 and issue #38. To verify my thought, I reverted the fix for issue #17 and updated the existing fragment dependent tests in the branch issue38. I built the LLVM project using the repo toolchain (the issue38 branch), both issue #17 and issue #38 have been fixed.
